src/harvest.js

```diff
@@ -0,0 +1,213 @@
+* License, or (at your option) any later version.
+*
+* melinda-record-import-harvester-publication-archives is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU Affero General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+* @licend  The above is the entire license notice
+* for the JavaScript code in this file.
+*
+*/
+
+import fs from 'fs';
+import {URL, URLSearchParams} from 'url';
+import moment from 'moment';
+import fetch from 'node-fetch';
+import HttpStatusCodes from 'http-status-codes';
+import nodeUtils from 'util';
+import {Utils} from '@natlibfi/melinda-commons';
+
+import xml2js from 'xml2js';
+import filterxml from 'filterxml';
+
+const {createLogger} = Utils;
+
+export default async function ({recordsCallback, harvestURL, harvestMetadata, harvestFilter, harvestFilterISBN, harvestFilterNamespace, pollInterval, pollChangeTimestamp, changeTimestampFile, failedHarvestFile, onlyOnce = false}) {
+	const logger = createLogger();
+	const parser = new xml2js.Parser();
+	let originalUrl = '';
+	let failedQueries = 0;
+
+	return process();
+
+	async function process({pollChangeTime} = {}) {
+		const setTimeoutPromise = nodeUtils.promisify(setTimeout);
+
+		pollChangeTime = pollChangeTime || getPollChangeTime();
+
+		const timeBeforeFetching = moment();
+
+		logger.log('debug', `Fetching records updated between ${pollChangeTime.format()} - ${timeBeforeFetching.format()}`);
+		try {
+			await harvest(null);
+		} catch (e) {
+			fs.writeFileSync(failedHarvestFile, e); // If failure happens on n:th cycle (resumption) previous cycles (n-1) are already sent. Save data of harvest to file just in case used later.
+			throw (e);
+		}
+
+		if (!onlyOnce) {
+			logger.log('debug', `Waiting ${pollInterval / 1000} seconds before polling again`);
+			await setTimeoutPromise(pollInterval);
+			writePollChangeTimestamp(timeBeforeFetching);
+			return process({pollChangeTime: timeBeforeFetching.add(1, 'seconds')});
+		}
+
+		function getPollChangeTime() {
+			if (fs.existsSync(changeTimestampFile)) {
+				const data = JSON.parse(fs.readFileSync(changeTimestampFile, 'utf8'));
+				return moment(data.timestamp);
+			}
+
+			if (pollChangeTimestamp) {
+				return moment(pollChangeTimestamp);
+			}
+
+			return moment();
+		}
+
+		function writePollChangeTimestamp(time) {
+			fs.writeFileSync(changeTimestampFile, JSON.stringify({
+				timestamp: time.format()
+			}));
+		}
+
+		// ListRecords can only fetch 100 records at the time.
+		// Each cycle fetches max 100 records, filters records, passes them to callback and passes possible resumption token to new cycle
+		async function harvest(token) {
+			const url = new URL(harvestURL);
+
+			if (token) {
```

Comment:
`URL`-instances have `searchParams`-property which can be used for setting the params (See [Helmet harvester](https://github.com/NatLibFi/melinda-record-import-harvester-helmet/blob/a7d8b6d30f92a7c8582e07edb2398e035502c0d3/src/harvest.js#L110) for reference).